### DISCUSSION TITLE
Merging / Combining Depth passes

### DISCUSSION THREAD

**Discussion by swingingtom** [pmndrs/postprocessing#328] [Q&A]
Hi postprocessing community,
Im wandering if there is a way to combine or merge depth passes ?

I need multiple scene render with different selection sets of objects. One in this passes stack is also masked.

Is there a way I could combine each render depth pass into one texture in order to assign this depth texture on the final pass such as it is done in https://codesandbox.io/s/qxqff ( @vanruesc as this is your demo, could you update this codesandox with the missing dependency ? I already forked, but it could be nice for others. This demo was the perfect starting point in my case 👍  )

```
const renderPass = new RenderPass(scene, camera);
const effectPass = new EffectPass(camera, smaaEffect, godRaysEffect);

// Use a custom depth texture.
const depthPass = new DepthPass(depthScene, camera);
effectPass.setDepthTexture(depthPass.renderTarget.texture, RGBADepthPacking);

```

As the time of writing this question, Im wondering if multiple render pass that only ClearColor would simply do it ?

The multiple render passes im talking about are sorted from farest to nearest, so an overwrite would not be an issue I guess.

**Comment by vanruesc** [comment]
Hi, I've updated the sandbox.

Using multiple depth passes is possible but the trick is to make them render to the same internal frame buffer. Unfortunately, this isn't very intuitive at the moment: if there's a pass that swaps buffers in between two depth passes, the second one will render to the wrong buffer. This type of setup will become easier and more predictable when explicit buffer management gets implemented.

**Comment by swingingtom** [comment]
Thanks!

> more predictable when explicit buffer management gets implemented
Out of curiosity, do you have any roadmap or document where I could already read about what this API will be? I found another post where you talk about it https://github.com/vanruesc/postprocessing/issues/229#issuecomment-706162586 but no real informations there.

Anyways, thanks for the fast reply ! I think I can follow your instructions, and if it breaks, I know that I need to check inbetween passes for swapsBuffer. 👍

**Comment by vanruesc** [comment]
> Out of curiosity, do you have any roadmap or document where I could already read about what this API will be?

There are no concrete concepts yet, but it's safe to assume that the concept of passes and effects will remain. Passes and effects will likely share the same API to specify input and output buffers, but built-in stuff will still configure this out of the box. The important part is that the user will be able to see and change the exact pipeline and all involved buffers without losing automatic functionality like SSAO's and DoF's access to a shared depth texture.

With the new API you'll be able to declare a specific buffer as output for the depth passes and as input for other passes or effects.

**Reply by netpro2k** [reply]
Is this still the plan? I have a similar situation where I want to render some objects after some effects passes have already applied but I want to respect depth of previous render passes. 

Currently my solution is just to do an additional copy/swap (via a empty **EffectPass** with `skipRendering=false` and `needsSwap=true`) to ensure both render passes end up using the same **RenterTarget**, and then I have the second not clear depth. This is obviously not ideal because there is an unnecessary copy (and is sort of abusing the current **EffectPass** in a hacky way). What I would ideally like to do is just tell the two render passes to use the same depth attachment.

**Reply by vanruesc** [reply]
> Is this still the plan?

Yes, I'm currently writing an overview of the planned redesign for a feedback/discussion ticket.
